src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java

```diff
@@ -114,27 +117,81 @@ private void annotatedEquivalence(TypeName type) {
-        + " Entry<java.lang.Byte, java.lang.Byte>";
-    AnnotationSpec typeUseAnnotation = AnnotationSpec.builder(TypeUseAnnotation.class).build();
     TypeName type = ParameterizedTypeName.get(Map.Entry.class, Byte.class, Byte.class)
-        .annotated(typeUseAnnotation);
-    String actual = type.toString();
-    assertEquals(expected, actual);
+        .annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString())
+        .isEqualTo("java.util.Map.@" + TUA + " Entry<java.lang.Byte, java.lang.Byte>");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedArrayType() {
+    TypeName type = ArrayTypeName.of(ClassName.get(Object.class)).annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString()).isEqualTo("java.lang.Object @" + TUA + " []");
+  }
+
+  @Test public void annotatedArrayElementType() {
+    TypeName type = ArrayTypeName.of(ClassName.get(Object.class).annotated(TYPE_USE_ANNOTATION));
+    assertThat(type.toString()).isEqualTo("@" + TUA + " java.lang.Object[]");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedOuterMultidimensionalArrayType() {
+    TypeName type = ArrayTypeName.of(ArrayTypeName.of(ClassName.get(Object.class)))
+        .annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString()).isEqualTo("java.lang.Object @" + TUA + " [][]");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedInnerMultidimensionalArrayType() {
+    TypeName type = ArrayTypeName.of(ArrayTypeName.of(ClassName.get(Object.class))
+        .annotated(TYPE_USE_ANNOTATION));
+    assertThat(type.toString()).isEqualTo("java.lang.Object[] @" + TUA + " []");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedArrayTypeVarargsParameter() {
+    TypeName type = ArrayTypeName.of(ArrayTypeName.of(ClassName.get(Object.class))
+        .annotated(TYPE_USE_ANNOTATION));
+    MethodSpec varargsMethod = MethodSpec.methodBuilder("m")
+        .addParameter(
+            ParameterSpec.builder(type, "p")
+                .build())
+        .varargs()
+        .build();
+    assertThat(varargsMethod.toString()).isEqualTo(""
+        + "void m(\n"
+        + "    java.lang.Object @" + TUA + " []... p) {\n"
+        + "}\n");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedArrayTypeInVarargsParameter() {
+    TypeName type = ArrayTypeName.of(ArrayTypeName.of(ClassName.get(Object.class))
+        .annotated(TYPE_USE_ANNOTATION));
+    MethodSpec varargsMethod = MethodSpec.methodBuilder("m")
+        .addParameter(
+            ParameterSpec.builder(type, "p")
+                .build())
+        .varargs()
+        .build();
+    assertThat(varargsMethod.toString()).isEqualTo(""
+        + "void m(\n"
+        + "    java.lang.Object[] @" + TUA + " ... p) {\n"
```

Comment:
whoa

Reply:
The other issue is that implementing these semantics requires either distinguishing between `TYPE_USE` annotations when annotating a type or change existing behavior which is a breaking change.

IMHO I like the simplicity of having implicit `TYPE_USE` semantics when annotating a type as the old usage is supported through ParameterSpec, MethodSpec, etc. I plan to make a PR with the implementation. Does this look as an acceptable solution or would you prefer to make the annotations aware whether they are `TYPE_USE` or not and emit them accordingly

Reply:
Yup, that makes sense. Good idea!

Reply:
I am not sure which approach you prefer. I assume that by yes you prefer the second choice, i.e. making annotations aware and emit accordingly, right?

Reply:
Actually I’d prefer that annotations on TypeName objects are always interpreted as type annotations.

That matches the language model, and it's flexible.